### PR TITLE
[tune/train] Consolidate checkpoint manager 2: Ray Train

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This is a follow-up from #24771 which moves the Ray Train implementation to use the new common checkpoint manager class.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by amogkam — COMMENTED** [review]
Thanks @krfricke! Overall, this PR looks good to me. But I want to make sure we're aligned on the end state of all this work.

Is this understanding correct?
1. The user facing API will eventually be `ray.ml.Checkpoint` for both Train and Tune, rather than directories for tune and dicts for train. This is what @xwjiang2010 is working on currently.
2.  After we have 1, Train and Tune can use a single common `CheckpointManager`, rather than each having its own subclass.
3. We want to eventually remove this code for HFTrainer: https://github.com/ray-project/ray/pull/24772/files#diff-4c32668d94aee848c135fb86eb4765c63b8671e157af7df881ed31a403a85b07R56-R88. This should be handled internally by the checkpoint manager rather than special cased for the Huggingface Trainer.

**Comment by krfricke** [comment]
That's correct @amogkam 

1. We may keep backwards compatibility with dicts/directories for now. I'll check with @xwjiang2010 to make sure
2. Yes, we want to only keep one checkpoint manager in the future
3. Yes, we should generally remove this or consolidate (currently tending to removing). I'm just keeping it here for now as a direct migration path

File: python/ray/air/train/integrations/huggingface/huggingface_trainer.py [hunk 1]
```diff
@@ -53,39 +57,54 @@
 # TODO(ml-team): Make dir syncing checkpoint logic generic.
 
 
-# The checkpoint is turned into a dict with node ip & path
-# in HuggingFaceTrainer.as_trainable
-# TODO(team-ml): Refactor checkpoint management along with Tune.
```
Comment (amogkam) [inline python/ray/air/train/integrations/huggingface/huggingface_trainer.py:58]: Leave this TODO?

File: python/ray/train/__init__.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 from ray.train.backend import BackendConfig
 from ray.train.callbacks import TrainingCallback
-from ray.train.checkpoint import CheckpointStrategy
+from ray.train._checkpoint import CheckpointStrategy
```
Comment (amogkam) [inline python/ray/train/__init__.py:3]:
```suggestion
from ray.utils.ml_utils import CheckpointStrategy
```

File: python/ray/train/trainer.py [hunk 3]
```diff
@@ -22,7 +22,7 @@
     construct_train_func,
     ActorWrapper,
 )
-from ray.train.checkpoint import (
+from ray.train._checkpoint import (
```
Comment (amogkam) [inline python/ray/train/trainer.py:25]: Let's import CheckpointStrategy from ml_utils?

**Review by amogkam — APPROVED** [review]
Thanks @krfricke!

**Comment by amogkam** [comment]
Actually we may want to move the common classes to ray.air right now itself. We’ll be deprecating ml_utils soon.

**Comment by krfricke** [comment]
thanks for the review - deprecating ML utils sounds good, I just want to make sure we're clear about where this code goes then. In `ray.air` it may be confusing with the `Checkpoint` class. I think one question would be if this should just go into Tune once we removed the custom implementations (just like `tune.report()`). 

Let's defer this when we actually deprecate the ml_utils - I'll attend to the other comments and merge once tests are passing.

**Pull request merged**